src/runtime/runtime.ts

```diff
@@ -54,25 +54,313 @@ export class Interner {
+                      GlobalInterner.intern(n), transaction, round, count);
+  }
+
+  toString() {
+    return `Change(${GlobalInterner.reverse(this.e)}, ${GlobalInterner.reverse(this.a)}, ${GlobalInterner.reverse(this.v)}, ${GlobalInterner.reverse(this.n)}, ${this.transaction}, ${this.round}, ${this.count})`;
+  }
+}
+
+type ChangeSet = Change[];
+
+/**
+ * A register is just a numerical offset into the solved prefix.
+ * We can't make this a type alias because we wouldn't be able to
+ * tell the difference between static numbers and registers in scans.
+ */
+
+class Register {
+  constructor(public offset:number) {}
+}
+
+function isRegister(x: any): x is Register {
+  return x && x.constructor === Register;
+}
+
+/** The ignore register is a sentinel value for ScanFields that tell the scan to completely ignore that field. */
+type IgnoreRegister = null;
+let IGNORE_REG:IgnoreRegister = null;
+
+/** A scan field may contain a register, a static interned value, or the IGNORE_REG sentinel value. */
+type ScanField = Register|ID|IgnoreRegister;
+
+type ResolvedEAVN = {e:ID|undefined|IgnoreRegister, a:ID|undefined|IgnoreRegister, v:ID|undefined|IgnoreRegister, n:ID|undefined|IgnoreRegister};
 
+/**
+ * A scan maps a set of bound variables to unbound variables.
+ */
+
+class Scan {
+  constructor(public e:ScanField,
+              public a:ScanField,
+              public v:ScanField,
+              public n:ScanField) {}
+
+  protected resolved:ResolvedEAVN = {e: undefined, a: undefined, v:undefined, n: undefined};
+
+  /**
+   * Resolve each scan field. The resolved object may contain one of three possible value types:
+   * - IGNORE_REG -- this field is entirely ignored by the scan.
+   * - undefined -- this field is a register that hasn't been filled in yet. We'll fill it if possible.
+   * - ID -- this field contains a static or already solved interned value.
+   */
+  resolve(prefix:ID[]) {
+    let resolved = this.resolved;
+    if(isRegister(this.e)) {
+      resolved.e = prefix[this.e.offset];
+    } else {
+      resolved.e = this.e;
+    }
 
+    if(isRegister(this.a)) {
+      resolved.a = prefix[this.a.offset];
+    } else {
+      resolved.a = this.a;
+    }
+
+    if(isRegister(this.v)) {
+      resolved.v = prefix[this.v.offset];
+    } else {
+      resolved.v = this.v;
+    }
+
+    if(isRegister(this.n)) {
+      resolved.n = prefix[this.n.offset];
+    } else {
+      resolved.n = this.n;
+    }
+
+    return resolved;
+  }
+
+  /**
+   * A field is unresolved if it is completely ignored by the scan or is an output of the scan.
+   */
+  fieldUnresolved(resolved:ResolvedEAVN, key: keyof ResolvedEAVN) {
+    return resolved[key] === IGNORE_REG || resolved[key] === undefined;
+  }
+
+  /**
+   * A field is not a static match if it is ignored, not a static field, or the input value does not match the static value.
+   */
+  notStaticMatch(input:Change, key: "e"|"a"|"v"|"n") {
```

Comment:
Disregard my first comment, I see that it's already done in the branch: https://github.com/witheve/Eve/pull/744/files#diff-677ffa07af969f1cb122adf871ee9fc8R88